### DISCUSSION TITLE
cropping before and after homography transformation

### DISCUSSION THREAD

**Discussion by qiangwang57** [kornia/kornia#1111] [Questions]
Good afternoon,

I have a task to locate and register a small image in a large image (~100 times larger). What I did was first to manually crop a patch in the large image, which are slightly larger than the small image in terms of FOV. Then I used homogrpahy regression for the registration of the small image with the patch. The result is very good. However, there are black area at the edge of registered images because of the nature of the registration. However, the padded black area should contain meaningful information since the original patch was cropped from a large image. What I want to do is to "recover" the lost information during the registration. 

Here is what I tried: used the homography matrix obtained from the small image and the patch to transform the large image and the coordinates of the patch. Then I have the warped large image (retain the size) and the new coordinates, and do the crop. But the result is not as expected. The first image below is the warped patch and the second is the cropped image from the warped large image.
![image](https://user-images.githubusercontent.com/49925149/122417463-678da900-cf81-11eb-9868-0afe76b6fe8a.png)

and the code is here:
```
def homographywaper(image, homo_matrix, target_h=256, target_w=256, inverse=False):
    image_tensor = image2tensor(image, (target_h, target_w))
    homo_matrix_tensor = torch.Tensor(homo_matrix)
    homo_matrix_tensor = torch.unsqueeze(homo_matrix_tensor, dim=0)
    if inverse:
        homo_matrix_tensor = torch.inverse(homo_matrix_tensor)
    warper = HomographyWarper(target_h, target_w)
    image_tensor_warped = warper(image_tensor, homo_matrix_tensor)
    image_warped = (tensor2image(image_tensor_warped, image_tensor_warped.shape[:2])*255).astype(np.uint8)
    return image_warped

image_warped = homographywaper(image, data["homo_matrix"], image.shape[0], image.shape[1], inverse=False)
src_start_pos = np.append(data["histo_patch_start_pos"], 1)
dst_start_pos = np.matmul(homo_matrix_tensor_denorm.squeeze(0).detach().cpu().numpy(), src_start_pos)
dst_start_pos /= dst_start_pos[-1]
src_end_pos = np.append(data["histo_patch_start_pos"] + data["patch_dim"], 1)
dst_end_pos = np.matmul(homo_matrix_tensor_denorm.squeeze(0).detach().cpu().numpy(), src_end_pos)
image_warped_patch = image_warped[int(dst_start_pos[1]):int(dst_end_pos[1]), int(dst_start_pos[0]):int(dst_end_pos[0])]
image_patch = image[int(src_start_pos[1]):int(src_start_pos[1])+data["patch_dim"], int(src_start_pos[0]):int(src_start_pos[0])+data["patch_dim"]]
image_patch_warped = cv2_warp(image_patch, data["homo_matrix"], image_patch.shape[0], image_patch.shape[1], inverse=False)

plt.subplot(121)
plt.imshow(image_patch_warped, cmap="gray")
plt.title("image patch warped")
plt.subplot(122)
plt.imshow(image_warped_patch, cmap="gray")
plt.title("image warped patch")
```

Any tips would be greatly appreciated!

Cheers, Chiang

**Comment by edgarriba** [comment]
@qiangwang57 is this question related too ?

**Reply by qiangwang57** [reply]
> @qiangwang57 is this question related too ?

@edgarriba this is another question that really makes me struggle. Any help would be appreciated!

**Reply by edgarriba** [reply]
can you provide the homography matrix and image to reproduce in a colab ? a possible error is that the homography_warper expects a normalized homography being the coordinates mapping in the range of [-1,1]

**Reply by qiangwang57** [reply]
> can you provide the homography matrix and image to reproduce in a colab ? a possible error is that the homography_warper expects a normalized homography being the coordinates mapping in the range of [-1,1]

Hi @edgarriba , thanks for the response. I have made a min code to reproduce the issue. The link is as https://colab.research.google.com/drive/1YNAcYQZ2UWRAuVCZgjj-iHi4dvIyQJgo#scrollTo=LSz3Ib3EyQ4w

Appreciate your help!

Cheers, Qiang

**Reply by qiangwang57** [reply]
> can you provide the homography matrix and image to reproduce in a colab ? a possible error is that the homography_warper expects a normalized homography being the coordinates mapping in the range of [-1,1]

@edgarriba I noticed that the original image does not exist in the Colab. I just upload it to somewhere else so that you can download it and add it to the Colab. The links is: https://ibb.co/7gMTmQc

**Comment by edgarriba** [comment]
@qiangwang57 we provide a new API for [`ImageRegistration`](https://kornia.readthedocs.io/en/latest/geometry.transform.html#kornia.geometry.transform.image_registrator.ImageRegistrator) - to avoid issues you might want to try or even check the low levels details to catch the error.

https://kornia.readthedocs.io/en/latest/applications/image_registration.html

**Comment by zhaoguangyuan123** [comment]
same issue here.
"crop and homography" and "homography and crop" are essentially different here. 

Hope there is concrete solution for it.

**Reply by edgarriba** [reply]
mind sharing a reproducible example ?

**Reply by zhaoguangyuan123** [reply]
https://colab.research.google.com/drive/1Hg0Lzc5FHp05ySDGcg36sHhM83xZOkSu?usp=sharing


See this link. 
I have created a clear example. 

Assume the homography matrix I used derived by homography estimation on the cropped images. How can it correctly be applied to the larger image before cropping?

Best,
Guangyuan

**Reply by zhaoguangyuan123** [reply]
Hi Edgar, 

And also how to properly dealing with homography matrix with scaling in kornia is another question.
Below is another Colab Notebook. If I want to apply homography onto a scaled image, how should I process the homography matrix?

https://colab.research.google.com/drive/1UaulwPCjMRCjN6lJgjNt29cF8PQIin7L?usp=sharing

Thanks,
Guangyuan

**Reply by zhaoguangyuan123** [reply]
Hi Edgar,

Pls kindly let me know if you do not have time to reply in recent days. Then I may look into details of your source code by myself. 
Best,
Guangyuan

**Reply by edgarriba** [reply]
to do crops i suggest to use `get_perspective_transform`: https://kornia.readthedocs.io/en/latest/geometry.transform.html#kornia.geometry.transform.get_perspective_transform

in your code i don't understand the inverse in H_scaled, while working with homographies i suggest to explicitly write the variables include from/to reference frame to avoid bugs. This is the most common error to have an inverse or place a multiplication in the wrong place.

check this answer: https://github.com/kornia/kornia/discussions/2045#discussioncomment-4272517

**Reply by zhaoguangyuan123** [reply]
> to do crops i suggest to use `get_perspective_transform`: https://kornia.readthedocs.io/en/latest/geometry.transform.html#kornia.geometry.transform.get_perspective_transform
> 
> in your code i don't understand the inverse in H_scaled, while working with homographies i suggest to explicitly write the variables include from/to reference frame to avoid bugs. This is the most common error to have an inverse or place a multiplication in the wrong place.
> 
> check this answer: [#2045 (comment)](https://github.com/kornia/kornia/discussions/2045#discussioncomment-4272517)

Thanks, I have modified the names in the two colab notebooks accordingly.

I will look into the function you are referring. 

Best,
Guangyuan

**Reply by zhaoguangyuan123** [reply]
> to do crops i suggest to use `get_perspective_transform`: https://kornia.readthedocs.io/en/latest/geometry.transform.html#kornia.geometry.transform.get_perspective_transform
> 
> in your code i don't understand the inverse in H_scaled, while working with homographies i suggest to explicitly write the variables include from/to reference frame to avoid bugs. This is the most common error to have an inverse or place a multiplication in the wrong place.
> 
> check this answer: [#2045 (comment)](https://github.com/kornia/kornia/discussions/2045#discussioncomment-4272517)


Sry I looked into the function and I might not get your point. 
That function "get_perspective_transform" seems to be useful for estimating the homography matrix while what I want to do is to use the pre-estimated the homography matrix to do the transform.

**Reply by edgarriba** [reply]
check how is used with crop augmentations https://github.com/kornia/kornia/blob/master/kornia/augmentation/_2d/geometric/crop.py#L139

**Comment by zhaoguangyuan123** [comment]
TBH, the document regarding the "**normlazation**" in kornia homography is not unclear and without so many examples.

**Reply by edgarriba** [reply]
ill check if i have time this week, this is not my main work. Notice that we are community project, not a company product.
Consider donating to pay staff to improve or contribute yourself.

https://opencollective.com/kornia

**Reply by zhaoguangyuan123** [reply]
> ill check if i have time this week, this is not my main work. Notice that we are community project, not a company product. Consider donating to pay staff to improve or contribute yourself.
> 
> https://opencollective.com/kornia

Thank you. Just donate some through the link. Really appreciate the repo that you and the team created.

**Comment by zhaoguangyuan123** [comment]
I have solved the issue I raised. This issue raised because I was not familiar with the normalized and un-normalized homography and also some basics about the cropping and scaling in homography. 

Suggest it would be better if Kornia include some basic tutorials on it as many users are not familiar with homography operations. 

Thanks,
Guangyuan

**Reply by edgarriba** [reply]
@zhaoguangyuan123 probably you are the one right one after this experience.
I'd be very happy to guide and support you in this task. /cc @ducha-aiki 

feel free to join slack and discuss in the `#general` channel
